luarules/gadgets/unit_shield_behaviour.lua

```diff
@@ -0,0 +1,350 @@
+function gadget:GetInfo()
+	return {
+		name = "Shield Behaviour",
+		desc = "Overrides default shield engine behavior. Defines downtime.",
+		author = "SethDGamre",
+		layer = 1,
+		enabled = true
+	}
+end
+
+if not Spring.GetModOptions().shieldsrework then return false end
+if not gadgetHandler:IsSyncedCode() then return end
+
+---- Optional unit customParams ----
+-- shield_downtime = <number in seconds>, if not set defaults to 5 seconds
+-- shield_aoe_penetration = bool, if true then AOE damage will hurt units within the shield radius
+
+local spGetUnitShieldState = Spring.GetUnitShieldState
+local spSetUnitShieldState = Spring.SetUnitShieldState
+local spGetGameSeconds = Spring.GetGameSeconds
+local spSetUnitShieldRechargeDelay = Spring.SetUnitShieldRechargeDelay
+local spDeleteProjectile = Spring.DeleteProjectile
+local spGetProjectileDefID = Spring.GetProjectileDefID
+local spGetUnitPosition = Spring.GetUnitPosition
+local spGetUnitsInSphere = Spring.GetUnitsInSphere
+local spGetProjectilesInRectangle = Spring.GetProjectilesInRectangle
+local spGetUnitIsStunned = Spring.GetUnitIsStunned
+
+local shieldUnitDefs = {}
+local shieldUnitsData = {}
+local originalShieldDamages = {}
+local flameWeapons = {}
+local unitDefIDCache = {}
+local projectileDefIDCache = {}
+local shieldedUnits = {}
+local AOEWeaponDefIDs = {}
+local projectileShieldHitCache = {}
+local gameSeconds = 0
+
+local allyTeams = Spring.GetAllyTeamList()
+local allTeams = {}
+
+for i, allyID in ipairs(allyTeams) do
+	local teamList = Spring.GetTeamList(allyID)
+
+	if teamList then
+		allyTeams[i] = teamList
+
+		for _, teamID in pairs(teamList) do
+			allTeams[teamID] = allyID
+		end
+	end
+end
+
+for weaponDefID, weaponDef in ipairs(WeaponDefs) do
+	local areaOfEffect = weaponDef.damageAreaOfEffect
+	local interceptedByShieldType = weaponDef.interceptedByShieldType
+
+	if areaOfEffect > 11 and not weaponDef.customParams.shield_aoe_penetration and weaponDef.interceptedByShieldType == 1 then -- 11 because the the benchmark cortex sumo has a AOE of 12
```

Comment:
Hardcoding the value negates the functionality of the bitmasks, meaning the definition in the weapon defs will not match what actually happens. This needs to be calculated, rather than compared to a static value.

Reply:
It's gonna be a rabbit hole you might want to dive into later. 

I was unable to find where the projectile types were assigned bitmasks. The spring wiki has some defaults defined but my testing didnt seem consistent with what the wiki said at the same.

Shield mysteries.

Reply:
The bitmasks are in the weapon defs, the [`interceptedByShieldType`](https://springrts.com/wiki/Gamedev:WeaponDefs#interceptedByShieldType) tag. (You can ignore the default values in the wiki, we're overriding all of that.)